include/seqan3/alignment/matrix/detail/trace_iterator_base.hpp

```diff
@@ -0,0 +1,295 @@
+class trace_iterator_base
+{
+private:
+
+    static_assert(std::same_as<value_type_t<matrix_iter_t>, trace_directions>,
+                  "Value type of the underlying iterator must be seqan3::detail::trace_directions.");
+
+    //!\brief Befriend with corresponding const_iterator.
+    template <typename other_derived_t, typename other_matrix_iter_t>
+    //!\cond
+        requires std::constructible_from<derived_t, other_derived_t> &&
+                 std::constructible_from<matrix_iter_t, other_matrix_iter_t>
+    //!\endcond
+    friend class trace_iterator_base;
+
+    //!\brief Befriend the derived iterator class to allow calling the private constructors.
+    friend derived_t;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr trace_iterator_base() = default; //!< Defaulted.
+    constexpr trace_iterator_base(trace_iterator_base const &) = default; //!< Defaulted.
+    constexpr trace_iterator_base(trace_iterator_base &&) = default; //!< Defaulted.
+    constexpr trace_iterator_base & operator=(trace_iterator_base const &) = default; //!< Defaulted.
+    constexpr trace_iterator_base & operator=(trace_iterator_base &&) = default; //!< Defaulted.
+    ~trace_iterator_base() = default; //!< Defaulted.
+
+    /*!\brief Constructs from the underlying trace matrix iterator indicating the start of the trace path.
+     * \param[in] matrix_iter The underlying matrix iterator.
+     */
+    constexpr trace_iterator_base(matrix_iter_t const matrix_iter) noexcept : matrix_iter{matrix_iter}
+    {
+        set_trace_direction(*matrix_iter);
+    }
+
+    /*!\brief Constructs from the underlying trace matrix iterator indicating the start of the trace path.
+     * \tparam other_matrix_iter_t The underlying matrix iterator type of `other`; the condition
+     *                             `std::constructible_from<matrix_iter_t, other_matrix_iter_t>` must evaluate to
+     *                             `true`.
+     * \param[in] other The underlying matrix iterator.
+     *
+     * \details
+     *
+     * Allows the conversion of non-const to const iterator.
+     */
+    template <typename other_derived_t, two_dimensional_matrix_iterator other_matrix_iter_t>
+    //!\cond
+        requires std::constructible_from<matrix_iter_t, other_matrix_iter_t>
+    //!\endcond
+    constexpr trace_iterator_base(trace_iterator_base<other_derived_t, other_matrix_iter_t> const & other) noexcept :
+        trace_iterator_base{other.matrix_iter}
+    {}
+    //!\}
+
+public:
+    /*!\name Associated types
+     * \{
+     */
+    using value_type = trace_directions; //!< The value type.
+    using reference = value_type; //!< The reference type.
+    using pointer = value_type *; //!< The pointer type.
+    using difference_type = std::ptrdiff_t; //!< The difference type.
+    using iterator_category = std::forward_iterator_tag; //!< Forward iterator tag.
```

Comment:
If the reference == value_type the category can at most be input. But the iterator_concept can be more. (I know this is also not correct in other places).